streams/src/main/java/org/apache/kafka/streams/processor/internals/CompositeRestoreListener.java

```diff
@@ -0,0 +1,118 @@
+
+class CompositeRestoreListener implements BatchingStateRestoreCallback, StateRestoreListener {
+
+    private static final NoOpStateRestoreListener NO_OP_STATE_RESTORE_LISTENER = new NoOpStateRestoreListener();
+    private final BatchingStateRestoreCallback internalBatchingRestoreCallback;
+    private final StateRestoreListener storeRestoreListener;
+    private StateRestoreListener globalRestoreListener = NO_OP_STATE_RESTORE_LISTENER;
+
+    CompositeRestoreListener(StateRestoreCallback stateRestoreCallback) {
+
+        if (stateRestoreCallback instanceof StateRestoreListener) {
+            storeRestoreListener = (StateRestoreListener) stateRestoreCallback;
+        } else {
+            storeRestoreListener = new NoOpStateRestoreListener();
+        }
+
+        internalBatchingRestoreCallback =
+            (BatchingStateRestoreCallback) ((stateRestoreCallback instanceof BatchingStateRestoreCallback)
+                                            ? stateRestoreCallback
+                                            : new WrappedBatchingStateRestoreCallback(stateRestoreCallback));
+    }
+
+    @Override
+    public void onRestoreStart(TopicPartition topicPartition, String storeName,
+                               long startingOffset, long endingOffset) {
+        globalRestoreListener.onRestoreStart(topicPartition, storeName, startingOffset, endingOffset);
+        storeRestoreListener.onRestoreStart(topicPartition, storeName, startingOffset, endingOffset);
+    }
+
+    @Override
+    public void onBatchRestored(TopicPartition topicPartition, String storeName,
+                                long batchEndOffset, long numRestored) {
+        globalRestoreListener.onBatchRestored(topicPartition, storeName, batchEndOffset, numRestored);
+        storeRestoreListener.onBatchRestored(topicPartition, storeName, batchEndOffset, numRestored);
+    }
+
+    @Override
+    public void onRestoreEnd(TopicPartition topicPartition, String storeName,
+                             long totalRestored) {
+        globalRestoreListener.onRestoreEnd(topicPartition, storeName, totalRestored);
+        storeRestoreListener.onRestoreEnd(topicPartition, storeName, totalRestored);
+
+    }
+
+    @Override
+    public void restoreAll(final Collection<KeyValue<byte[], byte[]>> records) {
+        internalBatchingRestoreCallback.restoreAll(records);
+    }
+
+    void setGlobalRestoreListener(StateRestoreListener globalRestoreListener) {
+        if (globalRestoreListener != null) {
+            this.globalRestoreListener = globalRestoreListener;
+        }
+    }
+
+    @Override
+    public void restore(final byte[] key, final byte[] value) {
+        throw new UnsupportedOperationException("Single restore functionality shouldn't be called directly but "
+                                                + "through the delegated StateRestoreCallback instance");
+    }
+
+
+    private static final class NoOpStateRestoreListener implements StateRestoreListener {
```

Comment:
Will still have one no-op method, but I guess it's worth it as it does reduce the boilerplate some.